### PR TITLE
AC - created new terraform modules

### PR DESCRIPTION
**Pull request by andrew-cleveland**
## Description

This introduces the use of modules into the galleri terraform, this is a way to template out a series of terraform resources with a sensible set of defaults which can be overwritten when required.

## Context

The primary purpose of this is to enable developers to create new resources quickly and easily.
The secondary purpose is to make it easier to understand what infrastructure is being created, but breaking down the environment into blocks

## Type of changes

- [ x ] Refactoring (non-breaking change)
- [ x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ x ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ x ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ x ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: scripts/pipeline/nonprod_lsoa_load/nonprod_lsoa_load.py [hunk 1]
```diff
@@ -100,6 +100,7 @@ def batch_write_to_dynamodb(lsoa_data):
     # repeat till no records left
     dynamodb_client = boto3.client('dynamodb')
     for i in range(1, 100000, 100):
+        print(i)
```
Comment (zainm97) [inline scripts/pipeline/nonprod_lsoa_load/nonprod_lsoa_load.py:103]: can remove

File: terraform/environment/dev-old/resources/lambda/clinicIcbList/lambdaHandler/clinicIcbListLambda.zip
Comment (zainm97) [inline terraform/environment/dev-old/resources/lambda/clinicIcbList/lambdaHandler/clinicIcbListLambda.zip:1]: we dont want to commit the zip files to the repo. These are just an artifact of doing a tf apply locally

File: terraform/environment/dev-old/resources/lambda/clinicInformation/lambdaHandler/clinicInformationLambda.zip
Comment (zainm97) [inline terraform/environment/dev-old/resources/lambda/clinicInformation/lambdaHandler/clinicInformationLambda.zip:1]: we dont want to commit the zip files to the repo. These are just an artifact of doing a tf apply locally

File: terraform/environment/dev-old/resources/lambda/clinicSummaryList/lambdaHandler/clinicSummaryListLambda.zip
Comment (zainm97) [inline terraform/environment/dev-old/resources/lambda/clinicSummaryList/lambdaHandler/clinicSummaryListLambda.zip:1]: we dont want to commit the zip files to the repo. These are just an artifact of doing a tf apply locally

File: terraform/environment/dev-old/resources/lambda/filterData/lambdaHandler/dataFilterLambda.zip
Comment (zainm97) [inline terraform/environment/dev-old/resources/lambda/filterData/lambdaHandler/dataFilterLambda.zip:1]: we dont want to commit the zip files to the repo. These are just an artifact of doing a tf apply locally

File: terraform/environment/dev-old/resources/lambda/invitationParameters/lambdaHandler/invitationParametersLambda.zip
Comment (zainm97) [inline terraform/environment/dev-old/resources/lambda/invitationParameters/lambdaHandler/invitationParametersLambda.zip:1]: we dont want to commit the zip files to the repo. These are just an artifact of doing a tf apply locally

File: terraform/environment/dev-old/resources/lambda/invitationParametersPutForecastUptake/lambdaHandler/invitationParametersPutForecastUptakeLambda.zip
Comment (zainm97) [inline terraform/environment/dev-old/resources/lambda/invitationParametersPutForecastUptake/lambdaHandler/invitationParametersPutForecastUptakeLambda.zip:1]: we dont want to commit the zip files to the repo. These are just an artifact of doing a tf apply locally

File: terraform/environment/dev-old/resources/lambda/invitationParametersPutQuintiles/lambdaHandler/invitationParametersPutQuintilesLambda.zip
Comment (zainm97) [inline terraform/environment/dev-old/resources/lambda/invitationParametersPutQuintiles/lambdaHandler/invitationParametersPutQuintilesLambda.zip:1]: we dont want to commit the zip files to the repo. These are just an artifact of doing a tf apply locally

File: terraform/main.tf [hunk 2]
```diff
@@ -0,0 +1,705 @@
+terraform {
+  backend "s3" {
+    bucket = "galleri-github-oidc-tf-aws-tfstates"
+    key    = "infra.tfstate"
+    region = "eu-west-2"
+    # dynamodb_table = "galleri-terraform-state"
```
Comment (zainm97) [inline terraform/main.tf:6]: do we need these in

File: terraform/main.tf [hunk 3]
```diff
@@ -0,0 +1,705 @@
...
+  source               = "./modules/cloudwatch"
+  lambda_function_name = module.clinic_summary_list_lambda.lambda_function_name
+  retention_days       = 14
+}
+
+module "clinic_summary_list_api_gateway" {
+  source                                = "./modules/api-gateway"
+  lambda_invoke_arn                     = module.clinic_summary_list_lambda.lambda_invoke_arn
+  api_gateway_path_part                 = "clinic-summary-list"
+  api_gateway_method_request_parameters = {}
```
Comment (zainm97) [inline terraform/main.tf:274]: clinic summary list takes "method.request.querystring.participatingIcb" = true as param

File: terraform/modules/api-gateway-deployment/variables.tf [hunk 4]
```diff
@@ -0,0 +1,3 @@
+variable "rest_api_id" {}
+variable "stage_name" {}
+# variable "depends_on" {}
```
Comment (zainm97) [inline terraform/modules/api-gateway-deployment/variables.tf:3]: do we need

File: terraform/modules/api-gateway/api-gateway.tf [hunk 5]
```diff
@@ -0,0 +1,105 @@
+resource "aws_api_gateway_rest_api" "galleri" {
+  name        = "galleri-dev-local-2"
+  description = "API for the galleri webapp"
+  endpoint_configuration {
+    types = ["REGIONAL"]
+  }
+}
+
+// CLINIC INFORMATION
```
Comment (zainm97) [inline terraform/modules/api-gateway/api-gateway.tf:9]: can remove comment if this is the generic

File: terraform/modules/api-gateway/variables.tf [hunk 6]
```diff
@@ -0,0 +1,10 @@
+variable "lambda_invoke_arn" {}
```
Comment (zainm97) [inline terraform/modules/api-gateway/variables.tf:1]: if there is a variable aws_api_gateway_integration for we can default the  integration_http_method = "POST" . This is the tricky bit that we couldnt figure but it actually means Post the lambda to the api gateway

File: terraform/main.tf (hunk 2 above)
Comment (andrew-cleveland) [reply terraform/main.tf:6]: Yes, in the future I will implement these so I have just left them commented out until everyone is on the same version

File: terraform/modules/api-gateway-deployment/variables.tf (hunk 4 above)
Comment (andrew-cleveland) [reply terraform/modules/api-gateway-deployment/variables.tf:3]: I want to set this in the future to fix the bug with the API gateway starting out of sequence so I will leave it in for now and raise a bug to address it.

**Review by msandhuNHS — DISMISSED** [review]

**Review by zainm97 — DISMISSED** [review]
looks good!

**Review by zainm97 — APPROVED** [review]

**Review by msandhuNHS — APPROVED** [review]

**Pull request merged**